src/NVMLDevicePool.cpp

```diff
@@ -0,0 +1,465 @@
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        return result;
+    }
+
+    double NVMLDevicePoolImp::utilization_mem(int accel_idx) const
+    {
+        check_accel_range(accel_idx);
+        double result = NAN;
+        nvmlUtilization_t util;
+        nvmlReturn_t nvml_result;
+
+        nvml_result =  nvmlDeviceGetUtilizationRates(m_nvml_device.at(accel_idx), &util);
+        if (nvml_result == NVML_SUCCESS) {
+            result = (double) util.memory;
+        }
+        else {
+            throw Exception("NVMLDevicePool::" + std::string(__func__) +
+                            ": NVML failed to get memory utilization for accelerator " +
+                            std::to_string(accel_idx) + ".  Error: " + nvmlErrorString(nvml_result),
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        return result;
+    }
+
+    std::vector<int> NVMLDevicePoolImp::active_process_list(int accel_idx) const
+    {
+        check_accel_range(accel_idx);
+        std::vector<int> result;
+        unsigned int temp = m_max_contexts;
+
+        nvmlReturn_t nvml_result;
+        nvmlProcessInfo_t *process_info_list;
+        process_info_list = new nvmlProcessInfo_t[temp];
+
+        nvml_result = nvmlDeviceGetComputeRunningProcesses(m_nvml_device[accel_idx], &temp, &process_info_list[0]);
+        if (nvml_result == NVML_SUCCESS) {
+            for (int i = 0; i<temp; i++) {
+                result.push_back(process_info_list[i].pid);
+            }
+        }
+        else {
+            if (nvml_result == NVML_ERROR_INSUFFICIENT_SIZE) {
+                process_info_list = new nvmlProcessInfo_t[temp];
+                nvml_result = nvmlDeviceGetComputeRunningProcesses(m_nvml_device[accel_idx], &temp, &process_info_list[0]);
+
+                if (nvml_result == NVML_SUCCESS) {
+                    for (int i = 0; i<temp; i++) {
+                        result.push_back(process_info_list[i].pid);
+                    }
+                }
+                else {
+                    throw Exception("NVMLDevicePool::" + std::string(__func__) +
+                                    ": NVML failed to acquire running processes for accelerator " +
+                                    std::to_string(accel_idx) + ".  Error: " + nvmlErrorString(nvml_result),
+                                    GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+                }
+            }
+            else {
+                throw Exception("NVMLDevicePool::" + std::string(__func__) +
+                                ": NVML failed to acquire running processes for accelerator " +
+                                std::to_string(accel_idx) + ".  Error: " + nvmlErrorString(nvml_result),
+                                GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+            }
+        }
+        return result;
+    }
+
+
+    void NVMLDevicePoolImp::frequency_control(int accel_idx, int setting) const
```

Comment:
Change the API to match what the NVML spec is.
```(int accel_idx, int min_freq, int max_freq)```

Also rename to frequency_control_sm.